tools/ttysnoop.py

```diff
@@ -62,33 +69,75 @@ def usage():
 bpf_text = """
 #include <uapi/linux/ptrace.h>
 #include <linux/fs.h>
+#include <linux/uio.h>
 
-#define BUFSIZE 256
+#define BUFSIZE USER_DATASIZE
 struct data_t {
     int count;
     char buf[BUFSIZE];
 };
 
+BPF_ARRAY(data_map, struct data_t, 1);
 BPF_PERF_OUTPUT(events);
 
+static int do_tty_write(void *ctx, const char __user *buf, size_t count)
+{
+    int zero = 0, i;
+    struct data_t *data;
+
+    data = data_map.lookup(&zero);
+    if (!data)
+        return 0;
+
+    for (i = 0; i < USER_DATACOUNT; i++) {
+        bpf_probe_read_user(&data->buf, BUFSIZE, (void *)buf);
```

Comment:
I did not see `buf` pointer change. That means, every iteration, you try to read the same source. Is this what you want? Also, we can use bpf_probe_read_user() return value which should tells you precisely the bytes read.

Reply:
ugh, yep.. fixed ;-)